File: 分治算法/offer-17.cpp
#include <vector>
#include <string>
#include <iostream>
using namespace std;
class Solution
{
public:
    vector<int> res;
    vector<int> printNumbers(int n)
    {
        string num(n, '0');
        dfs(0,n,num);
        return res;
    }

    void dfs(int x, int n, string &num)
    {
        if (x == n)
        {
            int ptr = 0;
            while(ptr<n&&num[ptr]=='0') ptr++;
            if(ptr!=n)
                res.push_back(stoi(num.substr(ptr)));
            return;
        }
        for (int i=0; i < 10; i++)
        {
            num[x] = '0' + i;
            dfs(x + 1,n,num);
        }
    }
};

int main()
{
    int n = 2;
    Solution obj;
    vector<int>res = obj.printNumbers(n);
    for(auto num:res)
        cout<<num<<endl;
}

File: 分治算法/offer-17.py
from typing import List


class Solution:
    def _printNumbers(self, n: int) -> List[int]:
        return list(range(1, 10**n))

    def printNumbers(self, n: int) -> List[int]:
        '''全排列'''
        num = ['0' for _ in range(n)]
        res = []
        def dfs(x):
            if x==n:
                ptr = 0
                while(ptr<n and num[ptr]=='0'):
                    ptr += 1
                if(ptr!=n):
                    res.append(int("".join(num[ptr:])))
                return
            for i in range(10):
                num[x] = str(i)
                dfs(x+1)
        dfs(0)
        return res
n = 2
obj = Solution()
print(obj.printNumbers(n))
